tests/data/dataset_test.py

```diff
@@ -19,22 +19,24 @@ def setUp(self):
         self.vocab.add_token_to_namespace("sentence")
         self.vocab.add_token_to_namespace(".")
         self.token_indexer = {"tokens": SingleIdTokenIndexer()}
+        self.instances = self.get_instances()
         super(TestDataset, self).setUp()
 
     def test_instances_must_have_homogeneous_fields(self):
         instance1 = Instance({"tag": (LabelField(1, skip_indexing=True))})
         instance2 = Instance({"words": TextField([Token("hello")], {})})
         with pytest.raises(ConfigurationError):
-            _ = Dataset([instance1, instance2])
+            batch = Batch([instance1, instance2])
+            batch._check_types()  # pylint: disable=protected-access
```

Comment:
This isn't called in the constructor anymore?  Why not?

Reply:
I thought we talked about how this check wasn't really appropriate in e.g. multi-task learning? I moved it to a method so we could add it back later if we decided we wanted it in some cases.

Reply:
I think `Batch` actually makes this more clear.  We can handle batching together instances with different fields in the data iterator, and each batch should have a consistent set of fields.  Otherwise things totally break.  So, yeah, I'd move this check back to the constructor.

Reply:
done